design/ceph/multus-network.md

```diff
@@ -90,111 +90,106 @@ interface will be added to the host network namespace for all nodes that will ru
-a single instance of the holder DaemonSet. A possible (future) partial implementation may be
-possible by restarting the CSI plugin Pods when network interfaces are added or removed.
-
-**Known issue:** the Docker container runtime does not use Linux's native `/var/run/netns`
-directory. This mitigation is known to work on cri-o runtime but not Docker. Therefore, this feature
-will be disabled by default and enabled optionally by the
-`ROOK_CSI_MULTUS_USE_HOLDER_MOVER_PATTERN=true` variable in the Rook-Ceph operator's config.
-
-A previous version of the CSI proposal had the holder Pods creating "setup" and "teardown"
-Kubernetes Jobs for migrating/un-migrating the multus networks. This design was rejected since new
-Jobs wouldn't be able to be created if the Kubernetes namespace were in "Terminating" state.
+When the CSI-sidecars DaemonSet is updated, the `cephcsi` binary will have a new version and
+potentially new arguments as well. The new `cephcsi` binary is copied (from the `copy-binary`
+init container) in the Kubelet HostPath volume directory.
+If copying fails, the container will keep retrying until the operation succeeds.
+On the side of the CSI-plugin DaemonSet, the directory will be watched for changes and will
+accordingly kill and start the new binary process. If the new binary process fails to start, the
+container will retry until the operation succeeds.
+
+If a copied binary is left in the Kubelet HostPath volume directory after the CSI-sidecars DaemonSet
+pod is removed, space will remain occupied on the filesystem by the `cephcsi` binary (60 MB) and
+the arguments file (a few bytes). We should try to avoid leaving the binary on the host as
+much as possible. At least, we might always leave a single binary only since each binary is
+overwritten, same for the args file. In order to mitigate this we will rely on a PostHook to clean up the copied binary
+along with the arguments file. However, cleanup is not guaranteed to be successful due to the
+PostHook's nature.
+TODO: remove this section if we decide to remove the binary from the Kubelet HostPath volume
+directory directly from the CSI-plugin DaemonSet
+
+There is no potential race to be worried about, if the CSI-plugin DaemonSet starts before the
+CSI-sidecars DaemonSet, it will simply keep watching for the binary to be present in Kubelet
+HostPath volume directory. Once it is detected, the `cephcsi` process will be started.
+
+When a new node is added, both DaemonSets are started, and the
+process described above occurs on the node.
+
+Both DaemonSets should be deleted when the CSI driver components are removed and
+the termination should remove the `cephcsi` binary and the arguments file.
+TODO: remove last sentence if we decide to remove the binary from the Kubelet HostPath volume directory directly from the CSI-plugin DaemonSet
+
+The initial implementation of this design is **not** limited to supporting a single CephCluster with
```

Comment:
So we can support multiple CephClusters with multus now? The double negative "not limited" is not clear to me. :)